src/OrchardCore.Modules/OrchardCore.Autoroute/Routing/AutorouteRoute.cs

```diff
@@ -44,31 +49,57 @@ public VirtualPathData GetVirtualPath(VirtualPathContext context)
                         }
                     }
 
-                    return new VirtualPathData(_target, path );
+                    return new VirtualPathData(_target, path);
                 }
             }
+            
 
             return null;
         }
 
-        public Task RouteAsync(RouteContext context)
+        public async Task RouteAsync(RouteContext context)
         {
             var requestPath = context.HttpContext.Request.Path.Value;
 
-            string contentItemId;
-
-            if(_entries.TryGetContentItemId(requestPath, out contentItemId))
+            if (_entries.TryGetContentItemId(requestPath, out var contentItemId))
             {
-                context.RouteData.Values["area"] = "OrchardCore.Contents";
-                context.RouteData.Values["controller"] = "Item";
-                context.RouteData.Values["action"] = "Display";
-                context.RouteData.Values["contentItemId"] = contentItemId;
+                await EnsureRouteData(context, contentItemId);
+                await _target.RouteAsync(context);
+            }
+        }
 
-                context.RouteData.Routers.Add(_target);
-                return _target.RouteAsync(context);
+        private async Task<RouteValueDictionary> GetContentItemDisplayRoutes(HttpContext context, string contentItemId)
+        {
+            if (string.IsNullOrEmpty(contentItemId))
+                return null;
+            var contentManager = context.RequestServices.GetService<IContentManager>();
+            var contentItem = await contentManager.GetAsync(contentItemId);
```

Comment:
Can you debug the code to ensure that when the default action is hit then it already finds the content item from the cache and doesn't reload it from database?

Then we need to confirm there is no performance impact on both displaying a content item and rendering links.

Reply:
Sure, I checked in SQL Profiler

exec sp_executesql N'select  TOP 1 [Document].* from [Document] inner join [ContentItemIndex] on [ContentItemIndex].[DocumentId] = [Document].[Id]  where (([ContentItemIndex].[ContentItemId] = @p0) and ([ContentItemIndex].[Published] = @p1))',N'@p0 nvarchar(4000),@p1 bit',@p0=N'4by24ckrbhcx16d4p988brck4w',@p1=1


Database got hit only once in AutorouteRoute, then in ItemController ContentItem returned from YesSql cache